packages/material-ui-lab/src/Rating/Rating.js

```diff
@@ -0,0 +1,475 @@
+      percent = (right - event.pageX - ownerWindow(rootNode).pageXOffset) / (width * max);
+    } else {
+      percent = (event.pageX - left - ownerWindow(rootNode).pageXOffset) / (width * max);
+    }
+
+    let newHover = Math.ceil((max * percent) / precision) * precision;
+    newHover = clamp(newHover, precision, max);
+    setState(prev =>
+      prev.hover === newHover && prev.focus === newHover
+        ? prev
+        : {
+            hover: newHover,
+            focus: newHover,
+          },
+    );
+
+    setFocusVisible(false);
+
+    if (onChangeActive && hover !== newHover) {
+      onChangeActive(event, newHover);
+    }
+  };
+
+  const handleMouseLeave = event => {
+    if (onMouseLeave) {
+      onMouseLeave(event);
+    }
+
+    const newHover = -1;
+    setState({
+      hover: newHover,
+      focus: newHover,
+    });
+
+    if (onChangeActive && hover !== newHover) {
+      onChangeActive(event, newHover);
+    }
+  };
+
+  const handleChange = event => {
+    if (onChange) {
+      onChange(event, parseFloat(event.target.value));
+    }
+  };
+
+  const handleFocus = event => {
+    if (isFocusVisible(event)) {
+      setFocusVisible(true);
+    }
+
+    const newFocus = parseFloat(event.target.value);
+    setState(prev => ({
+      hover: prev.hover,
+      focus: newFocus,
+    }));
+
+    if (onChangeActive && focus !== newFocus) {
+      onChangeActive(event, newFocus);
+    }
+  };
+
+  const handleBlur = event => {
+    if (hover !== -1) {
+      return;
+    }
+
+    if (focusVisible !== false) {
+      setFocusVisible(false);
+      onBlurVisible();
+    }
+
+    const newFocus = -1;
+    setState(prev => ({
+      hover: prev.hover,
+      focus: newFocus,
+    }));
+
+    if (onChangeActive && focus !== newFocus) {
+      onChangeActive(event, newFocus);
+    }
+  };
+
+  const item = (propsItem, state) => {
+    const id = `${name}-${String(propsItem.value).replace('.', '-')}`;
+    const container = (
+      <IconContainerComponent
+        {...propsItem}
+        className={clsx(classes.icon, {
+          [classes.iconEmpty]: !state.filled,
+          [classes.iconFilled]: state.filled,
+          [classes.iconHover]: state.hover,
+          [classes.iconFocus]: state.focus,
+          [classes.iconActive]: state.active,
+        })}
+      >
+        {emptyIcon && !state.filled ? emptyIcon : icon}
+      </IconContainerComponent>
+    );
+
+    if (readOnly || disabled) {
+      return <React.Fragment key={propsItem.value}>{container}</React.Fragment>;
+    }
+
+    return (
+      <React.Fragment key={propsItem.value}>
+        <label className={classes.label} htmlFor={id}>
+          {container}
+          <span className={classes.visuallyhidden}>{getLabelText(propsItem.value)}</span>
+        </label>
+        <input
+          onFocus={handleFocus}
+          onBlur={handleBlur}
+          onChange={handleChange}
+          value={propsItem.value}
+          id={id}
+          type="radio"
```

Comment:
I doubt a radio makes much sense for a discrete, ordered distribution of values. Why do we need an `input` and not a `role="slider"` with the appropriate aria attributes

Reply:
From the implementations I could benchmark, the accessibility part can potentially be handled in many different ways (radio, select, input range, button). I went with https://www.w3.org/WAI/tutorials/forms/custom-controls/.

Reply:
But those are considered with form controls. Why is this an necessary feature? We usually went with "people don't read from the DOM anyway". What changed?

Reply:
I don't know. I couldn't find any clear distinguishable advantage between the different options. I went with the WAI recommendations without thinking twice or knowing better. I would hope we can change that in the future if we need to.